### PR TITLE
Optimize TileRDDReproject using mutable region reprojection

### PR DESCRIPTION
**Pull request by jpolchlo**
This PR attempts to improve TileRDDReproject's performance by eliminating the merge step.  This is accomplished by using a mutable version of RasterRegionReproject to continually burn reprojected tiles into the same buffers.

### CODE REVIEW

File: raster/src/main/scala/geotrellis/raster/reproject/RasterRegionReproject.scala [hunk 1]
```diff
@@ -21,40 +21,162 @@ trait RasterRegionReproject[T <: CellGrid] extends Serializable {
     * @param resampleMethod cell value resample method
     */
   def regionReproject(raster: Raster[T], src: CRS, dest: CRS, rasterExtent: RasterExtent, region: Polygon, resampleMethod: ResampleMethod): Raster[T]
+
+  def mutableRegionReproject(target: T, raster: Raster[T], src: CRS, dest: CRS, rasterExtent: RasterExtent, region: Polygon, resampleMethod: ResampleMethod): Unit
+
 }
 
 object RasterRegionReproject {
+  def rowCoords(destRegion: Polygon, destRasterExtent: RasterExtent, toSrcCrs: Transform): Int => (Array[Int], Array[Double], Array[Double]) = {
```
Comment (echeipesh) [inline raster/src/main/scala/geotrellis/raster/reproject/RasterRegionReproject.scala]: This should be made `private` so it can be tweaked if desired arrises

File: spark/src/main/scala/geotrellis/spark/reproject/TileRDDReproject.scala [hunk 2]
```diff
@@ -208,7 +202,21 @@ object TileRDDReproject {
     // val expectedCols = targetDataExtent.width / newMetadata.layout.cellSize.width
     // println(s"Expected width: $expectedCols")
```
Comment (echeipesh) [inline spark/src/main/scala/geotrellis/spark/reproject/TileRDDReproject.scala]: debug comments, not part of this PR but can go now.

File: raster/src/main/scala/geotrellis/raster/reproject/RasterRegionReproject.scala [hunk 3]
```diff
@@ -21,40 +21,162 @@ trait RasterRegionReproject[T <: CellGrid] extends Serializable {
...
+          cfor(0)(_ < xs.size, _ + 1){ s =>
+            buffer.set(pxs(s), i, resampler.resample(xs(s), ys(s)))
+          }
+        }
+      }
+    }
+  }
+
+  implicit val multibandInstance = new RasterRegionReproject[MultibandTile] {
+    def regionReproject(raster: Raster[MultibandTile], src: CRS, dest: CRS, rasterExtent: RasterExtent, region: Polygon, resampleMethod: ResampleMethod): Raster[MultibandTile] = {
```
Comment (echeipesh) [inline raster/src/main/scala/geotrellis/raster/reproject/RasterRegionReproject.scala]: The code duplication between `regionReproject` and `mutableRegionReproejct` should be taken care of. Probably the easiest thing is to delegate from the former to the latter once tiles are allocated.

Comment (echeipesh) [reply raster/src/main/scala/geotrellis/raster/reproject/RasterRegionReproject.scala]: Any thoughts on naming convention for `mutable` methods now that its used?

**Comment by jpolchlo** [comment]
For the record, test runs on EMR yielded the following timings on a 20 executor cluster.  For the baseline case:

Baseline case

And for the mutable-optimized case:

Optimized case

Slight, but not insignificant performance improvement.

File: raster/src/main/scala/geotrellis/raster/reproject/RasterRegionReproject.scala (hunk 3 above)
Comment (jpolchlo) [reply raster/src/main/scala/geotrellis/raster/reproject/RasterRegionReproject.scala]: I don't have significant thoughts on naming convention, sad to say.  This works well enough for me.  Do you have specific concerns about the nomenclature used here?

**Comment by echeipesh** [comment]
Weird that the partition count for base pyramid job doesn't match between two ingests.
